lightning/src/ln/network_graph.rs

```diff
@@ -0,0 +1,1609 @@
+//! The top-level network map tracking logic lives here.
+
+use secp256k1::key::PublicKey;
+use secp256k1::Secp256k1;
+use secp256k1;
+
+use bitcoin_hashes::sha256d::Hash as Sha256dHash;
+use bitcoin_hashes::Hash;
+use bitcoin::blockdata::script::Builder;
+use bitcoin::blockdata::opcodes;
+
+use chain::chaininterface::{ChainError, ChainWatchInterface};
+use ln::features::{ChannelFeatures, NodeFeatures};
+use ln::msgs::{DecodeError,ErrorAction,LightningError,RoutingMessageHandler,NetAddress};
+use ln::msgs;
+use util::ser::{Writeable, Readable, Writer, ReadableArgs};
+use util::logger::Logger;
+
+use std::cmp;
+use std::sync::{RwLock,Arc};
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::collections::BTreeMap;
+use std::collections::btree_map::Entry as BtreeEntry;
+use std;
+
+
+/// Receives network updates from peers to track view of the network.
+pub struct NetworkGraph {
+	secp_ctx: Secp256k1<secp256k1::VerifyOnly>,
+	/// Representation of the payment channel network
+	pub network_map: RwLock<NetworkMap>,
+	chain_monitor: Arc<ChainWatchInterface>,
+	full_syncs_requested: AtomicUsize,
+	logger: Arc<Logger>,
+}
+
+impl NetworkGraph {
+	/// Creates a new tracker of the actual state of the network of channels and nodes.
+	pub fn new(our_pubkey: PublicKey, chain_monitor: Arc<ChainWatchInterface>, logger: Arc<Logger>) -> Self {
+		let mut nodes = BTreeMap::new();
+		nodes.insert(our_pubkey.clone(), NodeInfo {
+			channels: Vec::new(),
+			lowest_inbound_channel_fee_base_msat: u32::max_value(),
```

Comment:
If these are going to be pub we need something other than default values. We need them to actually be Option<>s and any groups of fields which are always either set or unset together should be in structs.